tests/Game.class.test.js

```diff
@@ -0,0 +1,916 @@
+      const game2048 = new Game();
+
+      game2048.start();
+
+      const flatState = game2048.getState().flat();
+      const nonEmptyCells = flatState.filter((cell) => cell !== 0);
+
+      expect(nonEmptyCells).toHaveLength(2);
+
+      expect(
+        nonEmptyCells.every((cell) => [2, 4].includes(cell)),
+      ).toBe(true);
+    });
+
+    it('should generate the board with 4s less often than with 2s', () => {
+      const game2048 = new Game();
+
+      let statesCombined = [];
+      let twos = 0;
+      let fours = 0;
+
+      for (let i = 0; i < 100; i++) {
+        game2048.start();
+
+        const flatState = game2048.getState().flat();
+
+        statesCombined = [
+          ...statesCombined,
+          ...flatState,
+        ];
+      }
+
+      statesCombined.forEach((cell) => {
+        if (cell === 2) {
+          twos++;
+        } else if (cell === 4) {
+          fours++;
+        }
+      });
+
+      expect(fours < twos).toBe(true);
+    });
+
+    it('should generate the board with cells in random positions', () => {
+      const game2048 = new Game();
+
+      const cellPositions = new Map();
+
+      for (let i = 0; i < 100; i++) {
+        game2048.start();
+
+        const state = game2048.getState();
+
+        state.forEach((row, rowIndex) => {
+          row.forEach((cell, cellIndex) => {
+            if (cell !== 0) {
+              const key = `${rowIndex}${cellIndex}`;
+
+              if (cellPositions.has(key)) {
+                cellPositions.set(key, cellPositions.get(key) + 1);
+              } else {
+                cellPositions.set(key, 1);
+              }
+            }
+          });
+        });
+      }
+
+      expect(cellPositions.size >= 8).toBe(true);
+      expect(cellPositions.size <= 16).toBe(true);
+    });
+  });
+
+  describe('board movement methods', () => {
+    describe('moveLeft', () => {
+      it('should not move the board if the game is not started', () => {
+        const game2048 = new Game([
+          [8, 0, 0, 0],
+          [0, 16, 0, 0],
+          [0, 0, 32, 0],
+          [0, 0, 0, 64],
+        ]);
+
+        game2048.moveLeft();
+
+        const state = game2048.getState();
+
+        expect(state).toEqual([
+          [8, 0, 0, 0],
+          [0, 16, 0, 0],
+          [0, 0, 32, 0],
+          [0, 0, 0, 64],
+        ]);
+      });
+
+      it('should move all non-zero cells to the left', () => {
+        const initialState = [
+          [8, 0, 0, 0],
+          [0, 16, 0, 0],
+          [0, 0, 32, 0],
+          [0, 0, 0, 64],
+        ];
+
+        const game2048 = new Game(cloneState(initialState));
+
+        game2048.start();
+        game2048.moveLeft();
+
+        const state = game2048.getState();
+
+        state.forEach((row, rowIndex) => {
+          const cursor = initialState[rowIndex].find((cell) => cell !== 0);
+
+          expect(checkIsCorrectlyAligned(row, 'start', cursor)).toBe(true);
+        });
```

Comment:
just check the first element in each line 
you don't need `checkIsCorrectlyAligned`

Reply:
Using your approach:
1) initial:
```json
[
  [8, 0, 0, 0],
  [0, 16, 0, 0],
  [0, 0, 32, 0],
  [0, 0, 0, 64],
]
```

2) after `start()`:
```json
[
  [8, 0, 0, 0],
  [0, 16, 2, 0],
  [0, 0, 32, 0],
  [0, 0, 4, 64],
]
```

3) student's incorrect implementation in `moveLeft()`:
```json
[
  [8, 0, 0, 0],
  [16, 0, 2, 0],
  [32, 0, 2, 0],
  [4, 0, 0, 64],
]
```

4) test passed, but solution is incorrect